### ISSUE TITLE
Unable to get/set array on Balance Point screen using API

### ISSUE THREAD

**Issue by Nancy-Salpepi** [phetsims/mean-share-and-balance#345]
**Test device**
MacBook Air M1 chip

**Operating System**
14.5

**Browser**
Safari 17.5

**Problem description**
For https://github.com/phetsims/qa/issues/1134, using the code from the examples.md to get an array for the Balanced Point screen results in an error. I don't have this issue with the other screens and it also works in the rc.1 version. 

**Steps to reproduce**
1. In studio on the Balance Point screen, kick 3 balls
2. In the console add: `await phetioClient.invokeAsync( 'meanShareAndBalance.balancePointScreen.model', 'getDataPoints', [] );` to the 'top' or 'studio' level
 
**Visuals**
<img width="1162" alt="Screenshot 2024-08-14 at 4 31 16 PM" src="https://github.com/user-attachments/assets/3ebe1e94-3176-435b-86ee-72ba43b84090">

**Comment by Nancy-Salpepi** [comment]
I also get that error when trying to set the array using the snippet of code in examples.md.

**Comment by jbphet** [comment]
This may also affect center-and-variability, so we should verify that this feature is working correctly in that sim once this is fixed.

**Comment by jbphet** [comment]
@marlitas - The problem appears to be related to the implementation fields in `BalancePointModel.BalancePointModelIO`.  The patch below makes the problem go away, but I didn't understand the motivation behind the changes the introduced the problem enough to feel confident in committing this.  The problem has to do with the finer points of using the `apply` function.  Please review and apply if you think this is a reasonable solution.

```diff
Subject: [PATCH] fix assertion on phet-io API call, see https://github.com/phetsims/mean-share-and-balance/issues/345
---
Index: js/balance-point/model/BalancePointModel.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/js/balance-point/model/BalancePointModel.ts b/js/balance-point/model/BalancePointModel.ts
--- a/js/balance-point/model/BalancePointModel.ts	(revision 431ad8b359b09fd8c719b6c77c6f2f8d68ea352e)
+++ b/js/balance-point/model/BalancePointModel.ts	(date 1723762250373)
@@ -120,8 +120,8 @@
       setDataPoints: {
         returnType: VoidIO,
         parameterTypes: [ ArrayIO( NumberIO ) ],
-        implementation: function( model: BalancePointModel, dataPoints: number[] ) {
-          const selectedSceneModel = model.selectedSceneModelProperty.value;
+        implementation: function( this: BalancePointModel, dataPoints: number[] ) {
+          const selectedSceneModel = this.selectedSceneModelProperty.value;
           selectedSceneModel.setDataPoints( dataPoints );
 
           // When phet-io clients set the data points statically there should be no animation and therefore no queued kicks.
@@ -140,8 +140,10 @@
       getDataPoints: {
         returnType: ArrayIO( NumberIO ),
         parameterTypes: [],
-        implementation: ( model: BalancePointModel ) =>
-          model.selectedSceneModelProperty.value.getSortedStackedObjects().map( soccerBall => soccerBall.valueProperty.value ),
+        implementation: function( this: BalancePointModel ) {
+          return this.selectedSceneModelProperty.value
+            .getSortedStackedObjects().map( soccerBall => soccerBall.valueProperty.value );
+        },
         documentation: 'Gets the data points for the selected scene model.'
       }
     }
```

**Comment by marlitas** [comment]
@jbphet and I talked about this in person and he will investigate `apply()` in conjunction with arrow functions for a bit then commit the fix. 

Part of the problem was my fault in not regression testing this refactor properly.

**Comment by jbphet** [comment]
I spent a limited amount of time on this, and essentially concluded that arrow functions don't work well with `apply` (more on this below), and having `this` as the first parameter for the functions matters.  So, I went ahead and committed the patch above, and it does appear to fix the problem on main.

I checked the behavior on the main version of Center and Variability, and it appears to be working, and the implementation there (see `CAVModel.ts, currently lines 157-173) matches that of the commit I just made, so I feel this supports the approach taken.

A note on arrow functions and the `apply` method in the[ Mozilla docs on arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) I found a portion that says: 
> [T]he call, apply, and bind methods are not useful when called on arrow functions, because arrow functions establish this based on the scope the arrow function is defined within, and the this value does not change based on how the function is invoked.

This sentence is specifically about `this`, so I'm not sure that it completely validates the decision to avoid using arrow functions in this context, but it does seem like a piece of supporting evidence.

**Comment by jbphet** [comment]
I searched through the code base to see if there were any other implementations of IOType methods that had `implementation` fields that used the arrow function style and had parameters.  I only found them in one place - `SoccerSceneModel`.  I just discussed these with @marlitas over Zoom and she helped me come up with commands to invoke these methods.  Our testing showed that these methods were also broken.  They have now been fixed, and I referenced this issue in the commit.  The breaking change was made to these `method` fields after Center and Variability was published, so there is no need to do a maintenance release.
